presto-main/src/test/java/io/prestosql/execution/buffer/BenchmarkPagesSerde.java

```diff
@@ -0,0 +1,217 @@
+/*
```

Comment:
Please model benchmarks after `io/prestosql/orc/BenchmarkColumnReaders.java`. This means benchmarking different column serdes (we can start with `Long` block only). Block serialization consumes large chunk of time for parges serde.

Reply:
I actually started working on similar benchmarks, here is my draft:
```suggestion
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.prestosql.execution.buffer;

import com.google.common.collect.ImmutableList;
import io.airlift.slice.BasicSliceInput;
import io.airlift.slice.OutputStreamSliceOutput;
import io.airlift.slice.Slice;
import io.airlift.slice.Slices;
import io.prestosql.spi.Page;
import io.prestosql.spi.block.Block;
import io.prestosql.spi.block.BlockBuilder;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OperationsPerInvocation;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Warmup;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;
import org.openjdk.jmh.runner.options.VerboseMode;
import org.testng.annotations.Test;

import java.io.File;
import java.io.FileOutputStream;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Random;
import java.util.concurrent.TimeUnit;

import static com.google.common.io.Files.createTempDir;
import static io.prestosql.execution.buffer.PagesSerdeUtil.readPages;
import static io.prestosql.execution.buffer.PagesSerdeUtil.writePages;
import static io.prestosql.spi.type.BigintType.BIGINT;
import static java.nio.channels.FileChannel.MapMode.READ_ONLY;
import static java.nio.file.Files.readAllBytes;
import static java.util.UUID.randomUUID;
import static java.util.concurrent.TimeUnit.MILLISECONDS;

@State(Scope.Thread)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Fork(3)
@Warmup(iterations = 30, time = 500, timeUnit = MILLISECONDS)
@Measurement(iterations = 20, time = 500, timeUnit = MILLISECONDS)
@BenchmarkMode(Mode.AverageTime)
@OperationsPerInvocation(BenchmarkPagesSerde.ROWS)
public class BenchmarkPagesSerde
{
    public static final int ROWS = 10_000_000;
    private static final PagesSerde SERDE = new TestingPagesSerdeFactory(
            new TestingBlockEncodingSerde(encoding -> new LongArrayBlockToFixedWidthBlockEncoding()), false).createPagesSerde();

    @Benchmark
    public Object readLongWithNull(BenchmarkData data)
    {
        return ImmutableList.copyOf(readPages(SERDE, new BasicSliceInput(data.getDataSource())));
    }

    @State(Scope.Thread)
    public static class BenchmarkData
    {
        private final Random random = new Random(0);
        private File temporaryDirectory;
        private File file;
        private Slice dataSource;

        @Setup
        public void setup()
                throws Exception
        {
            temporaryDirectory = createTempDir();
            file = new File(temporaryDirectory, randomUUID().toString());
            BlockBuilder blockBuilder = BIGINT.createBlockBuilder(null, 1024);
            Iterator<?> values = createValues();
            while (values.hasNext()) {
                Object value = values.next();
                if (value == null) {
                    blockBuilder.appendNull();
                }
                else {
                    BIGINT.writeLong(blockBuilder, ((Number) value).longValue());
                }
            }
            Page page = new Page(blockBuilder.build());
            writePages(SERDE, new OutputStreamSliceOutput(new FileOutputStream(file)), page);
            dataSource = Slices.wrappedBuffer(readAllBytes(file.toPath()));
        }

        public Slice getDataSource()
        {
            return dataSource;
        }

        protected Iterator<?> createValues()
        {
            List<Long> values = new ArrayList<>();
            for (int i = 0; i < ROWS; ++i) {
                if (random.nextBoolean()) {
                    values.add(random.nextLong());
                }
                else {
                    values.add(null);
                }
            }
            return values.iterator();
        }
    }

    @Test
    public void testReadLongWithNull()
            throws Exception
    {
        BenchmarkPagesSerde benchmark = new BenchmarkPagesSerde();
        BenchmarkData data = new BenchmarkData();
        data.setup();
        benchmark.readLongWithNull(data);
    }

    public static void main(String[] args)
            throws Exception
    {
        Options options = new OptionsBuilder()
                .verbosity(VerboseMode.NORMAL)
                .include(".*" + BenchmarkPagesSerde.class.getSimpleName() + ".*")
                .build();

        new Runner(options).run();
    }
}
```

Reply:
I would rather address the broader problem of benchmarking the entire serialization / deserialization performance in a variety of block encodings and data types after this PR. In the context of this PR, it doesn't especially matter what the data type and block encoding performance is as long as it remains constant overhead between the baseline and with reused buffers.